File: ai/schemas.py
from typing import Literal
from pydantic import BaseModel


# --- Defense ---


class Strength(BaseModel):
    title: str
    explanation: str
    line_references: list[str]
    file_paths: list[str]  # e.g. ["src/allocator.c", "include/arena.h"]


class DefenseOutput(BaseModel):
    strengths: list[Strength]
    design_intent: str
    defense_summary: str


# --- Prosecution ---


class Concern(BaseModel):
    title: str
    severity: Literal["low", "medium", "high", "critical"]
    explanation: str
    line_references: list[str]
    file_paths: list[str]  # e.g. ["src/allocator.c", "include/arena.h"]


class ProsecutionOutput(BaseModel):
    concerns: list[Concern]
    attack_vectors: list[str]
    prosecution_summary: str


# --- Interrogator ---


class Question(BaseModel):
    id: str
    text: str
    topic: str
    expected_concepts: list[str]


class QuestionsOutput(BaseModel):
    items: list[Question]


# --- Judge ---


class VerdictOutput(BaseModel):
    risk_score: float
    recommendation: Literal["approve", "request_changes", "needs_understanding_proof"]
    summary: str


# --- Evaluator ---


class PerQuestionScore(BaseModel):
    question_id: str
    score: int
    feedback: str
    missing_concepts: list[str]


# --- API contracts ---


class TrialRequest(BaseModel):
    diff: str
    repo_full_name: str
    pr_number: int
    language: str


class TrialResponse(BaseModel):
    defense: DefenseOutput
    prosecution: ProsecutionOutput
    questions: QuestionsOutput
    verdict: VerdictOutput


class QuestionRef(BaseModel):
    id: str
    text: str
    expected_concepts: list[str]


class EvaluateRequest(BaseModel):
    questions: list[QuestionRef]
    answers: list[str]


class EvaluateResponse(BaseModel):
    understanding_score: int
    per_question: list[PerQuestionScore]
    passed: bool

File: web/src/lib/schemas.ts
import { z } from "zod";

// --- Defense ---

export const StrengthSchema = z.object({
  title: z.string(),
  explanation: z.string(),
  line_references: z.array(z.string()),
  file_paths: z.array(z.string()),
});

export const DefenseOutputSchema = z.object({
  strengths: z.array(StrengthSchema),
  design_intent: z.string(),
  defense_summary: z.string(),
});

// --- Prosecution ---

export const ConcernSchema = z.object({
  title: z.string(),
  severity: z.enum(["low", "medium", "high", "critical"]),
  explanation: z.string(),
  line_references: z.array(z.string()),
  file_paths: z.array(z.string()),
});

export const ProsecutionOutputSchema = z.object({
  concerns: z.array(ConcernSchema),
  attack_vectors: z.array(z.string()),
  prosecution_summary: z.string(),
});

// --- Interrogator ---

export const QuestionSchema = z.object({
  id: z.string(),
  text: z.string(),
  topic: z.string(),
  expected_concepts: z.array(z.string()),
});

export const QuestionsOutputSchema = z.object({
  items: z.array(QuestionSchema),
});

// --- Judge ---

export const VerdictOutputSchema = z.object({
  risk_score: z.number(),
  recommendation: z.enum(["approve", "request_changes", "needs_understanding_proof"]),
  summary: z.string(),
});

// --- Trial response ---

export const TrialResponseSchema = z.object({
  defense: DefenseOutputSchema,
  prosecution: ProsecutionOutputSchema,
  questions: QuestionsOutputSchema,
  verdict: VerdictOutputSchema,
});

// --- Evaluate ---

export const PerQuestionScoreSchema = z.object({
  question_id: z.string(),
  score: z.number().int(),
  feedback: z.string(),
  missing_concepts: z.array(z.string()),
});

export const EvaluateResponseSchema = z.object({
  understanding_score: z.number().int(),
  per_question: z.array(PerQuestionScoreSchema),
  passed: z.boolean(),
});

// --- Inferred types ---

export type Strength = z.infer<typeof StrengthSchema>;
export type DefenseOutput = z.infer<typeof DefenseOutputSchema>;
export type Concern = z.infer<typeof ConcernSchema>;
export type ProsecutionOutput = z.infer<typeof ProsecutionOutputSchema>;
export type Question = z.infer<typeof QuestionSchema>;
export type QuestionsOutput = z.infer<typeof QuestionsOutputSchema>;
export type VerdictOutput = z.infer<typeof VerdictOutputSchema>;
export type TrialResponse = z.infer<typeof TrialResponseSchema>;
export type PerQuestionScore = z.infer<typeof PerQuestionScoreSchema>;
export type EvaluateResponse = z.infer<typeof EvaluateResponseSchema>;
